### PR TITLE
feat: adding updating of `mainnet_revisions.json` to the repo

### PR DESCRIPTION
**Pull request by NikolaMilosa**
Previously we used to have this workflow in a different repository. With this we are syncing it to this place with all other related code.

### CODE REVIEW

File: .github/workflows/update-mainnet-revisions.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 50
+
+      - name: Update IC versions file
+        run: |
+          set -eEuxo pipefail
+
+          time bazel run //ci/src/mainnet_revisions:mainnet_revisions --action_env=PUSH_TO_IC=${{ steps.app-token.outputs.token }}
```
Comment (basvandijk) [inline .github/workflows/update-mainnet-revisions.yaml:34]: You probably want to enable to `ci` config to make it consistent with other bazel invocations on CI:
```suggestion
          time bazel run --config=ci //ci/src/mainnet_revisions:mainnet_revisions --action_env=PUSH_TO_IC=${{ steps.app-token.outputs.token }}
```

**Review by cgundy — APPROVED** [review]
Nice, thank you! 🙏

File: .github/workflows/update-mainnet-revisions.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        uses: actions/create-github-app-token@v1
+        id: app-token
+        with:
+          app-id: ${{ vars.PR_AUTOMATION_BOT_PUBLIC_APP_ID }}
+          private-key: ${{ secrets.PR_AUTOMATION_BOT_PUBLIC_PRIVATE_KEY }}
+
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 50
```
Comment (basvandijk) [inline .github/workflows/update-mainnet-revisions.yaml:28]: You probably want to persist the token in the local git config like:
```suggestion
          fetch-depth: 50
          token: ${{ steps.app-token.outputs.token }}
```
Then you don't need to do the `~/.git-credentials` business below.

File: ci/src/mainnet_revisions/mainnet_revisions.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    repo_root = pathlib.Path(get_repo_root())
+
+    running_on_ci = os.environ.get("CI") or os.environ.get("GITHUB_ACTION")
+    if running_on_ci:
+        logging.info("Setting up git credentials for CI")
+        subprocess.check_call(["git", "config", "--global", "credential.helper", "store --file ~/.git-credentials"])
+        with open(os.path.expanduser("~/.git-credentials"), "w") as f:
+            f.write(f"{remote_url}\nusername=oauth2\npassword={ic_repo_push_token}\n")
+    else:
+        logging.info("Running locally, using default git credentials of the current user")
```
Comment (basvandijk) [inline ci/src/mainnet_revisions/mainnet_revisions.py:126]: See my comment about persisting the token in the git config using actions/checkout above. With that I don't think you need this:
```suggestion
```

File: ci/src/mainnet_revisions/mainnet_revisions.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+    if not repo_root.exists():
+        subprocess.check_call(["git", "clone", "--depth=50", remote_url, repo_root])
+    else:
+        subprocess.check_call(["git", "fetch", "origin", "--force"], cwd=repo_root)
+        subprocess.check_call(["git", "checkout", "master"], cwd=repo_root)
+        subprocess.check_call(["git", "reset", "--hard", "origin/master"], cwd=repo_root)
```
Comment (basvandijk) [inline ci/src/mainnet_revisions/mainnet_revisions.py:133]: Why is this necessary if actions/checkout has already checked out the master branch?

Comment (NikolaMilosa) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: Because its bazel and everything from it is run in a sandbox environment. I could hack it with `--sandbox_add_mount_pair` but I would rather not.

Comment (basvandijk) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: Are you sure? `bazel build/test` will indeed run the build or test in a sandbox but IIUC `bazel run` will just build the binary and then run it in the calling environment.

Comment (basvandijk) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: Although according to https://bazel.build/docs/user-manual#running-executables the binary will run in the runfiles tree of the binary. But `BUILD_WORKSPACE_DIRECTORY` will point to the root of the workspace.

Comment (basvandijk) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: Actually `BUILD_WORKING_DIRECTORY` would be better to use since it points to the directory where bazel was invoked from which should be the repo you just checked out.

Comment (NikolaMilosa) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: My only issue with that is that if somebody is trying to run it from local they basically can't because it will checkout master and remove their work. Adding exceptions to that basically adds unneeded complexity. What do you think?

File: ci/src/mainnet_revisions/mainnet_revisions.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                    "create",
+                    "--head",
+                    branch,
+                    "--repo",
+                    repo,
+                    "--fill",
+                ],
+                cwd=repo_root,
+                env=gh_env,
+            )
```
Comment (basvandijk) [inline ci/src/mainnet_revisions/mainnet_revisions.py:174]: Note that the default behavior of subprocess is to inhert the current process’ environment so I think you can just remove setting `env`.

File: ci/src/mainnet_revisions/mainnet_revisions.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        subnet=app_subnet_id, version=current_app_subnet_version, repo_root=pathlib.Path(repo_root)
+    )
+    git_modified_files = subprocess.check_output(["git", "ls-files", "--modified", "--others"], cwd=repo_root).decode(
+        "utf8"
+    )
+    if SAVED_VERSIONS_PATH in git_modified_files:
+        logging.info("Creating/updating a MR that updates the saved NNS subnet revision")
+        subprocess.check_call(["git", "add", SAVED_VERSIONS_PATH], cwd=repo_root)
+        if running_on_ci:
+            subprocess.check_call(["git", "config", "user.email", "infra@dfinity.org"], cwd=repo_root)
```
Comment (basvandijk) [inline ci/src/mainnet_revisions/mainnet_revisions.py:143]: We use the following email address in [other places](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/dfinity/ic%24+user.email&patternType=keyword&sm=0):
```suggestion
            subprocess.check_call(["git", "config", "user.email", "infra+github-automation@dfinity.org"], cwd=repo_root)
```

File: ci/src/mainnet_revisions/mainnet_revisions.py (hunk 4 above)
Comment (basvandijk) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: Ok, I understand the concern that you might want to run this locally and don't want the script to modify your current branch.

Then I'll let you decide what to do. I'm fine either way I guess.

Comment (NikolaMilosa) [reply ci/src/mainnet_revisions/mainnet_revisions.py]: I decided to have a really easy running experience of just using: 
```bash
python ci/src/mainnet_revisions/mainnet_revisions.py
```

It does introduce a dep to public dashboard but the worst that can happen is that it times out so I guess its fine

**Review by basvandijk — APPROVED** [review]
